### PR TITLE
TypeScript: convert `HistoryProvider`

### LINKED ISSUE #12666: TypeScript: Convert history provider

### PR DESCRIPTION
**Pull request by miina**
## Context
Converts `historyProvider` into TypeScript

## Relevant Technical Choices
~Ignores `design-system` import for now. Is there a better option?~

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12666

### CODE REVIEW

File: packages/story-editor/src/app/history/historyProvider.tsx [hunk 1]
```diff
@@ -17,14 +17,16 @@
...
-import PropTypes from 'prop-types';
 import {
   useCallback,
   useEffect,
   useState,
   useRef,
 } from '@googleforcreators/react';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment -- @todo Let's fix as part of #12656.
+// @ts-ignore
 import { useGlobalKeyDownEffect } from '@googleforcreators/design-system';
```
Comment (miina) [inline packages/story-editor/src/app/history/historyProvider.tsx:26]: @barklund / @swissspidy `useGlobalKeyDownEffect` is part of a larger file which has a bunch of related functions. Is there a better alternative without just doing TS ignore for now and without having to convert all those?

I could alternatively leave this ignore in for now and create a new issue for converting the `keyboard` folder of the design-system which would also remove this ignore here.

Thoughts?

Comment (swissspidy) [reply packages/story-editor/src/app/history/historyProvider.tsx:26]: How much work would it be to convert it? It seems worth doing

Comment (miina) [reply packages/story-editor/src/app/history/historyProvider.tsx:26]: You mean worth doing it within this PR instead of separately? Probably a 2-pointer, at the first glance doesn't look like there's anything overly complex. I'll just add it within this PR then.

File: packages/design-system/src/components/keyboard/index.tsx [hunk 2]
```diff
@@ -55,50 +63,41 @@ const CLICKABLE_INPUT_TYPES = [
   'reset',
 ];
 
-const globalRef = createRef();
+const globalRef = createRef<HTMLElement>();
 
 function setGlobalRef() {
   if (!globalRef.current) {
     globalRef.current = document.documentElement;
```
Comment (miina) [inline packages/design-system/src/components/keyboard/index.tsx]: Not sure what to do with this error 🤔 
`Cannot assign to 'current' because it is a read-only property.`

File: packages/design-system/src/components/keyboard/index.tsx [hunk 3]
```diff
@@ -55,50 +63,41 @@ const CLICKABLE_INPUT_TYPES = [
...
-  deps = undefined
+  refOrNode: RefOrNode,
+  keyNameOrSpec: KeyNameOrSpec,
+  type: string | undefined,
+  callback: KeyEffectCallback,
+  deps: DependencyList
 ) {
   const { keys } = useContext(Context);
   //eslint-disable-next-line react-hooks/exhaustive-deps -- Pass through provided deps.
   const batchingCallback = useBatchingCallback(callback, deps || []);
```
Comment (miina) [inline packages/design-system/src/components/keyboard/index.tsx]: Looks like `useBatchingCallback` doesn't work with `(event) => ...` currently. Getting the following error:
```
Argument of type 'KeyEffectCallback' is not assignable to parameter of type '(...args: unknown[]) => unknown'.
  Types of parameters 'event' and 'args' are incompatible.
    Type 'unknown' is not assignable to type 'KeyboardEvent'.
```

Probably needs some adjustments there.

File: packages/design-system/src/components/keyboard/index.tsx (hunk 2 above)
Comment (spacedmonkey) [reply packages/design-system/src/components/keyboard/index.tsx]: https://stackoverflow.com/questions/73967855/cannot-assign-to-current-because-it-is-a-read-only-property

Comment (miina) [reply packages/design-system/src/components/keyboard/index.tsx]: `useRef` and `createRef` are different, doesn't work like this for `createRef`.

Maybe we could just use `let` instead of a ref here.

File: packages/story-editor/src/app/history/historyProvider.tsx [hunk 4]
```diff
@@ -17,14 +17,14 @@
...
  * External dependencies
  */
-import PropTypes from 'prop-types';
 import {
   useCallback,
   useEffect,
   useState,
   useRef,
 } from '@googleforcreators/react';
 import { useGlobalKeyDownEffect } from '@googleforcreators/design-system';
```
Comment (miina) [inline packages/story-editor/src/app/history/historyProvider.tsx:26]: @barklund / @swissspidy Any thoughts on why this error happens here:
`Cannot find module '@googleforcreators/design-system' or its corresponding type declarations.`

Not sure what I'm missing here 🤔 It was ok just before.

File: packages/react/src/useBatchingCallback.ts [hunk 5]
```diff
@@ -31,7 +31,10 @@ import type { DependencyList } from 'react';
  * @param [deps] The optional callback dependencies.
  * @return The memoized batching function.
  */
-function useBatchingCallback<T extends (...args: unknown[]) => unknown>(
+function useBatchingCallback<
+  T extends (...args: [event: KeyboardEvent]) => void
```
Comment (miina) [inline packages/react/src/useBatchingCallback.ts]: Not sure if that's the best solution here. LMK if you have any thoughts.

File: packages/story-editor/src/app/history/historyProvider.tsx (hunk 4 above)
Comment (miina) [reply packages/story-editor/src/app/history/historyProvider.tsx:26]: Looks like needed to include `index.ts` as well to the `tsconfig.json`, however, then all the untyped errors come out. Had to remove the `design-system` as a reference [2017585](https://github.com/GoogleForCreators/web-stories-wp/pull/12674/commits/201758538ff969e90b962ba6aa9d3b0abe0e4356), however, then the module is not found again 🤷

Comment (miina) [reply packages/story-editor/src/app/history/historyProvider.tsx:26]: Ok, reverted that last change.

Basically now the `keyboard` part is typed, `design-system` is in the general `tsconfig.json` and also in the `story-editor` package's config, however, the error is still there since the `index.ts` file is not included in the `design-system`'s `tsconfig` file (if we include it, all the errors show up).

Perhaps now we can use `ts-ignore` for the module not found issue in this file (until the `design-system` package gets typed since the relevant part is actually typed? Thoughts, @swissspidy?

Comment (spacedmonkey) [reply packages/story-editor/src/app/history/historyProvider.tsx:26]: Having the same problem [here](https://github.com/GoogleForCreators/web-stories-wp/actions/runs/3481863339/jobs/5823491322#step:10:9) as well.

Comment (swissspidy) [reply packages/story-editor/src/app/history/historyProvider.tsx:26]: @miina The issue is that the `types` field in the `design-system` `package.json` points to `dist-types/index.d.ts` but no such file exists yet since `index.ts` is not listed in `tsconfig.json` (which is fine).

In 68bd6bf3b97c63a287c2d624b91bd9ab08921279 I added a quick workaround for that which should resolve this.

Thanks for converting all the keyboard stuff by the way 🚀

File: packages/react/src/useBatchingCallback.ts (hunk 5 above)
Comment (barklund) [reply packages/react/src/useBatchingCallback.ts]: Shouldn't it just always return type `T`? Wouldn't that work?

```ts
function useBatchingCallback<T extends (...args: unknown[]) => void>(
  callback: T,
  deps: DependencyList
): T {
```

File: packages/story-editor/src/app/history/historyProvider.tsx [hunk 6]
```diff
@@ -33,7 +33,11 @@ import usePreventWindowUnload from '../../utils/usePreventWindowUnload';
 import useHistoryReducer from './useHistoryReducer';
 import Context from './context';
 
-function HistoryProvider({ children, size }) {
+interface HistoryProviderProps {
+  children: ReactNode;
+  size: number;
+}
```
Comment (barklund) [inline packages/story-editor/src/app/history/historyProvider.tsx]: Please use `PropsWithChildren<{ size?: number }>`. And `size` should be optional.

File: packages/story-editor/src/app/history/reducer.ts [hunk 7]
```diff
@@ -13,6 +13,16 @@
...
+ */
+import type {
+  ReducerState,
+  ReducerProps,
+  HistoryEntry,
+} from '../../types/historyProvider';
+
 export const SET_CURRENT_STATE = 'set_state';
 export const CLEAR_HISTORY = 'clear';
 export const REPLAY = 'replay';
```
Comment (barklund) [inline packages/story-editor/src/app/history/reducer.ts]: Change to enum

File: packages/story-editor/src/app/history/useHistory.ts [hunk 8]
```diff
@@ -21,9 +21,13 @@ import { identity, useContextSelector } from '@googleforcreators/react';
...
  */
+import type { HistoryState } from '../../types/historyProvider';
 import Context from './context';
 
-function useHistory(selector) {
+function useHistory(): HistoryState;
+function useHistory<T>(
+  selector: (state: HistoryState) => T | HistoryState = identity
+) {
   return useContextSelector(Context, selector ?? identity);
```
Comment (barklund) [inline packages/story-editor/src/app/history/useHistory.ts]: No need for `?? identity` now as it's the default value.

File: packages/story-editor/src/types/historyProvider.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+export interface Actions {
+  stateToHistory: (state: HistoryEntry) => void;
+  clearHistory: () => void;
+  resetNewChanges: () => void;
+  undo: (offset: number) => boolean;
+  redo: (offset: number) => boolean;
+}
+export interface State {
+  currentEntry: HistoryEntry;
```
Comment (barklund) [inline packages/story-editor/src/types/historyProvider.ts:69]: Does it always have a current entry or should this be allowed null/undefined?

File: packages/story-editor/src/types/historyProvider.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  SET_CURRENT_STATE,
+} from '../app/history/reducer';
+
+export interface HistoryEntry {
+  story: Story;
+  selection: string[];
+  current: string | null;
+  pages: Page[];
+  capabilities: Record<string, boolean>;
+}
```
Comment (barklund) [inline packages/story-editor/src/types/historyProvider.ts:33]: I would _really_ love for this whole type (the entry in the history object) to be generic with respect to the history provider, but that's probably not gonna happen. I originally intended for it to be generic at least, but it's so intertwined with the story model, that it's probably gonna stay that way for now.

File: packages/story-editor/src/utils/usePreventWindowUnload.ts [hunk 11]
```diff
@@ -57,7 +67,9 @@ function usePreventWindowUnload() {
   return setPreventUnload;
 }
 
+declare const WEB_STORIES_DISABLE_PREVENT: string;
 const shouldDisablePrevent =
   typeof WEB_STORIES_DISABLE_PREVENT !== 'undefined' &&
   WEB_STORIES_DISABLE_PREVENT === 'true';
+// eslint-disable-next-line @typescript-eslint/no-empty-function -- Needed here.
```
Comment (barklund) [inline packages/story-editor/src/utils/usePreventWindowUnload.ts]: Can't it just be `() => () => undefined`?

File: packages/design-system/src/components/keyboard/index.tsx [hunk 12]
```diff
@@ -200,84 +185,70 @@ export function useIsKeyPressed(refOrNode, keyNameOrSpec, deps = undefined) {
...
-      const node =
-        typeof refOrNode?.current !== 'undefined'
-          ? refOrNode.current
-          : refOrNode;
+      let nodeEl;
+      if (refOrNode && 'current' in refOrNode) {
+        nodeEl = refOrNode.current;
+      } else {
+        nodeEl = refOrNode;
+      }
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: Move to a tiny util function? Would also get rid of the `let`. `const nodeEl = getNode(refOrNode);`

File: packages/design-system/src/components/keyboard/index.tsx [hunk 13]
```diff
@@ -307,36 +278,46 @@ function resolveKeySpec(keyDict, keyNameOrSpec) {
...
     return keys;
   }
   return keys.concat(keys.map((key) => `shift+${key}`));
 }
 
+interface KeyHandlerProps {
+  repeat: boolean;
+  editable: boolean;
+  clickable: boolean;
+  dialog: boolean;
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: I think they're all optional?

File: packages/design-system/src/components/keyboard/index.tsx [hunk 14]
```diff
@@ -346,7 +327,7 @@ function createKeyHandler(
   };
 }
 
-function isClickableTarget({ tagName, type }) {
+function isClickableTarget({ tagName, type }: HTMLInputElement) {
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: But buttons and links aren't inputs though? The function could event be considered a type guard, so it could be something like:

```ts
type ClickableHTMLElement = HTMLInputElement | HTMLAnchorElement | HTMLButtonElement;
function isClickableTarget(element: HTMLElement): element is ClickableHTMLElement {
```

But perhaps that's overkill.

File: packages/design-system/src/components/keyboard/index.tsx [hunk 15]
```diff
@@ -356,7 +337,12 @@ function isClickableTarget({ tagName, type }) {
   return false;
 }
 
-function isEditableTarget({ tagName, isContentEditable, type, readOnly }) {
+function isEditableTarget({
+  tagName,
+  isContentEditable,
+  type,
+  readOnly,
+}: HTMLInputElement) {
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: Again, this should work for non-inputs as well due to the `contentEditable` part. Or at least for textarea's too.

File: packages/design-system/src/components/keyboard/index.tsx [hunk 16]
```diff
@@ -393,14 +377,11 @@ export function isPlatformMacOS() {
 
 /**
  * Get the key specific to operating system.
- *
- * @param {string} key The key to replace. Options: [alt, ctrl, mod, cmd, shift].
- * @return {string} the mapped key. Returns the argument if key is not in options.
  */
-export function getKeyForOS(key) {
+export function getKeyForOS(key: string): string {
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: Return type shouldn't be necessary?

File: packages/design-system/src/components/keyboard/index.tsx [hunk 17]
```diff
@@ -393,14 +377,11 @@ export function isPlatformMacOS() {
...
   const isMacOS = isPlatformMacOS();
 
-  const replacementKeyMap = {
+  const replacementKeyMap: Record<string, string> = {
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: Isn't this type inferred? But then you probably have to do `key in replacementKeyMap` below.

File: packages/design-system/src/components/keyboard/index.tsx [hunk 18]
```diff
@@ -498,17 +475,15 @@ export function createShortcutAriaLabel(shortcut) {
     .join(delimiter);
 }
 
-const Kbd = () => <kbd />;
+const Kbd = ({ children }: { children: ReactNode }) => <kbd>{children}</kbd>;
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: Please use `PropsWithChildren`.

File: packages/design-system/src/components/keyboard/index.tsx [hunk 19]
```diff
@@ -498,17 +475,15 @@ export function createShortcutAriaLabel(shortcut) {
...
- * @param {Object} props props
- * @param {import('react').Component} props.component Component used to render the shortcuts. Defaults to `<kbd/>`
- * @param {string} props.shortcut Keyboard shortcut combination, e.g. 'shift+mod+z'.
- * @return {Node} Prettified keyboard shortcut.
  */
-export function Shortcut({ component: Component = Kbd, shortcut = '' }) {
+export function Shortcut({
+  component: Component = Kbd,
+  shortcut = '',
+}): ReactNode {
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: Is the return type required here?

File: packages/design-system/src/components/keyboard/index.tsx [hunk 20]
```diff
@@ -18,19 +18,26 @@
  * External dependencies
  */
 import Mousetrap from 'mousetrap';
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: This is a 500+ line file now. Maybe split it up into a few files?

File: packages/design-system/src/components/keyboard/index.tsx [hunk 21]
```diff
@@ -291,9 +262,9 @@ function resolveKeySpec(keyDict, keyNameOrSpec) {
     dialog = false,
     allowDefault = false,
   } = keySpec;
-  const mappedKeys = []
+  const mappedKeys = ([] as string[])
```
Comment (barklund) [inline packages/design-system/src/components/keyboard/index.tsx]: Not a big fan of this because it's ugly. Maybe `new Array<string>()` is cleaner?

File: packages/design-system/src/types/keyboard.ts [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+export interface Keys {
+  undo: string;
+  redo: string;
+  delete: string[];
+  clone: string;
+}
+export type KeyEffectCallback = (event: KeyboardEvent) => void;
+export type KeyNameOrSpec = KeySpec | string | string[];
+
+export type RefOrNode = Element | RefObject<Element>;
```
Comment (barklund) [inline packages/design-system/src/types/keyboard.ts]: It can also be `undefined`, no? Don't we do that sometimes to avoid the handler being attached?

File: packages/design-system/src/types/keyboard.ts [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+  repeat?: boolean;
+  editable?: boolean;
+  allowDefault?: boolean;
+}
+export interface Keys {
+  undo: string;
+  redo: string;
+  delete: string[];
+  clone: string;
+}
```
Comment (barklund) [inline packages/design-system/src/types/keyboard.ts]: I think it's better to `import  { keys } from 'keys.js'` and do `type Keys = typeof keys` here.

File: packages/design-system/src/types/keyboard.ts [hunk 24]
```diff
@@ -0,0 +1,40 @@
+/*
```
Comment (barklund) [inline packages/design-system/src/types/keyboard.ts:1]: Shouldn't this file exist inside the keyboard folder? Are the types required outside of the design system package? And even if they are, I still think they should be defined inside the keyboard folder actually.

**Review by barklund — COMMENTED** [review]
A bunch of comments.

File: packages/design-system/src/components/keyboard/index.tsx (hunk 18 above)
Comment (barklund) [reply packages/design-system/src/components/keyboard/index.tsx]: Also, is this default ever used? It was clearly broken before, given that it ignore the passed-in children, so I guess it's never been used? If so, let's remove it.

File: packages/design-system/src/components/keyboard/index.tsx (hunk 14 above)
Comment (miina) [reply packages/design-system/src/components/keyboard/index.tsx]: > But buttons and links aren't inputs though?

Buttons actually are, links aren't indeed, didn't think about those. Or well, maybe button is only when using an `input` with `button` type?

File: packages/story-editor/src/app/history/reducer.ts (hunk 7 above)
Comment (miina) [reply packages/story-editor/src/app/history/reducer.ts]: Do you mean creating an enum type for these instead of using  `typeof X`? As it currently is, we won't have to keep the enum AND the constants both updated.
Also, when using enum with TypeScript, it won't work with the reducer the same way -- currently, TS understand that if it's `set_state` then specific params are expected, if it's `clear` then something else, etc. In case of `enum`, it doesn't seem to understand that. I'll double check again but at least the last time I tried, it didn't work like that.

File: packages/design-system/src/components/keyboard/index.tsx (hunk 20 above)
Comment (miina) [reply packages/design-system/src/components/keyboard/index.tsx]: We could leave it for the `design-system` conversion work perhaps.

File: packages/design-system/src/types/keyboard.ts (hunk 24 above)
Comment (miina) [reply packages/design-system/src/types/keyboard.ts:1]: So far we've added a `types.ts` file in the `src` folder. Since one `types.ts` would get too large then I've created a `types` folder instead. If there's a different preference, should we do the same in all packages perhaps?
To me it makes sense to have consistency.

Is there a best practice for this? Maybe let's create a follow-up issue then where we'd move all the non-exported types out of the `types.ts` files?

File: packages/story-editor/src/types/historyProvider.ts (hunk 10 above)
Comment (miina) [reply packages/story-editor/src/types/historyProvider.ts:33]: Yep, it's probably going to stay this way for now 😄

File: packages/story-editor/src/app/history/historyProvider.tsx (hunk 6 above)
Comment (miina) [reply packages/story-editor/src/app/history/historyProvider.tsx]: Didn't know `PropsWithChildren` existed, thanks!

File: packages/story-editor/src/types/historyProvider.ts (hunk 9 above)
Comment (miina) [reply packages/story-editor/src/types/historyProvider.ts:69]: I'll adjust in #12411 as it will be in use there.

File: packages/design-system/src/components/keyboard/index.tsx (hunk 17 above)
Comment (miina) [reply packages/design-system/src/components/keyboard/index.tsx]: It's inferred indeed and added it just for TS to understand that the index is a string.
It doesn't seem to work with `return key in replacementKeyMap ? replacementKeyMap[key] : key;`. Or were you thinking of something else?

File: packages/react/src/useBatchingCallback.ts (hunk 5 above)
Comment (miina) [reply packages/react/src/useBatchingCallback.ts]: Doesn't seem so, getting the following error then:
`Argument of type '(...args: unknown[]) => void' is not assignable to parameter of type 'T'.`

File: packages/design-system/src/types/keyboard.ts (hunk 22 above)
Comment (miina) [reply packages/design-system/src/types/keyboard.ts]: I think we pass `{ current: null }` or something like that if we avoid the handler. Found one case where we use `null` as well, will add that.

File: packages/story-editor/src/app/history/reducer.ts (hunk 7 above)
Comment (miina) [reply packages/story-editor/src/app/history/reducer.ts]: All ok now after specifying in Slack with Morten,

File: packages/design-system/src/components/keyboard/index.tsx (hunk 20 above)
Comment (miina) [reply packages/design-system/src/components/keyboard/index.tsx]: Moved utils into a separate file.

File: packages/design-system/src/theme/index.js [hunk 25]
```diff
@@ -17,7 +17,7 @@
 /**
  * External dependencies
  */
-import type { DefaultTheme } from 'styled-components';
+import { DefaultTheme } from 'styled-components';
```
Comment (spacedmonkey) [inline packages/design-system/src/theme/index.js:20]: This lint error seems valid. I think this needs to be two files. A index.ts and theme.js.

Comment (spacedmonkey) [inline packages/design-system/src/theme/index.js]: Let's not import this type, it is not needed.

File: packages/react/src/useBatchingCallback.ts (hunk 5 above)
Comment (barklund) [reply packages/react/src/useBatchingCallback.ts]: Hm, not sure why that doesn't work, but this does. I don't know if we ever need callbacks with more than one argument, but if we do, this supports up to three args (and it's butt-ugly):

```ts
function useBatchingCallback<T>(
  callback: (arg: T) => void,
  deps: DependencyList
): (arg: T) => void;
function useBatchingCallback<T, U>(
  callback: (arg1: T, arg2: U) => void,
  deps: DependencyList
): (arg1: T, arg2: U) => void;
function useBatchingCallback<T, U, V>(
  callback: (arg1: T, arg2: U, arg3: V) => void,
  deps: DependencyList
): (arg1: T, arg2: U, arg3: V) => void;
function useBatchingCallback<T extends (...args: unknown[]) => void>(
  callback: T,
  deps: DependencyList
) {
  return useCallback(
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument -- Just passing through
    (...args) => batchedUpdates(() => callback(...args)),
    // eslint-disable-next-line react-hooks/exhaustive-deps -- Pass through provided deps.
    deps
  );
}
```

File: packages/design-system/src/types/keyboard.ts (hunk 24 above)
Comment (barklund) [reply packages/design-system/src/types/keyboard.ts:1]: Sure, let's do that.

File: packages/story-editor/src/utils/usePreventWindowUnload.ts [hunk 26]
```diff
@@ -57,7 +67,10 @@ function usePreventWindowUnload() {
...
 }
 
+declare const WEB_STORIES_DISABLE_PREVENT: string;
 const shouldDisablePrevent =
   typeof WEB_STORIES_DISABLE_PREVENT !== 'undefined' &&
   WEB_STORIES_DISABLE_PREVENT === 'true';
-export default shouldDisablePrevent ? () => () => {} : usePreventWindowUnload;
+export default shouldDisablePrevent
+  ? () => () => undefined
+  : usePreventWindowUnload;
```
Comment (spacedmonkey) [inline packages/story-editor/src/utils/usePreventWindowUnload.ts:76]: Why is this needed?

**Review by barklund — APPROVED** [review]
Looks great, let's get this one in!

**Review by spacedmonkey — APPROVED** [review]
Looks good to me

File: packages/story-editor/src/utils/usePreventWindowUnload.ts (hunk 26 above)
Comment (miina) [reply packages/story-editor/src/utils/usePreventWindowUnload.ts:76]: Are you asking why is the change `{}` to `undefined` needed?
Since `{}` is not allowed by the linting rules then using `undefined` instead. Originally I ignored the eslint error but Morten had the idea of using `undefined` instead. It doesn't really change anything.

File: packages/design-system/src/components/keyboard/keyboard.tsx [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+  const batchingCallback = useBatchingCallback(callback, deps || []);
+  useEffect(
+    () => {
+      const nodeEl = getNodeFromRefOrNode(refOrNode);
+      if (!nodeEl) {
+        return undefined;
+      }
+      if (
+        nodeEl.nodeType !== /* ELEMENT */ 1 &&
+        nodeEl.nodeType !== /* DOCUMENT */ 9
```
Comment (swissspidy) [inline packages/design-system/src/components/keyboard/keyboard.tsx]: Assuming this works:

```suggestion
        nodeEl.nodeType !== Node.ELEMENT_NODE &&
        nodeEl.nodeType !== Node.DOCUMENT_NODE
```
